rclcpp/src/rclcpp/intra_process_manager.cpp

```diff
@@ -12,69 +12,153 @@
+
+  publishers_[id].publisher = publisher;
+  publishers_[id].topic_name = publisher->get_topic_name();
+  publishers_[id].qos = publisher->get_actual_qos().get_rmw_qos_profile();
+
+  // Initialize the subscriptions storage for this publisher.
+  pub_to_subs_[id] = SplittedSubscriptions();
+
+  // create an entry for the publisher id and populate with already existing subscriptions
+  for (auto & pair : subscriptions_) {
+    if (can_communicate(publishers_[id], pair.second)) {
+      insert_sub_id_for_pub(pair.first, id, pair.second.use_take_shared_method);
+    }
   }
+
   return id;
 }
 
 uint64_t
-IntraProcessManager::add_subscription(
-  rclcpp::SubscriptionBase::SharedPtr subscription)
+IntraProcessManager::add_subscription(SubscriptionIntraProcessBase::SharedPtr subscription)
 {
+  std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
   auto id = IntraProcessManager::get_next_unique_id();
-  impl_->add_subscription(id, subscription);
+
+  subscriptions_[id].subscription = subscription;
+  subscriptions_[id].topic_name = subscription->get_topic_name();
+  subscriptions_[id].qos = subscription->get_actual_qos();
+  subscriptions_[id].use_take_shared_method = subscription->use_take_shared_method();
+
+  // adds the subscription id to all the matchable publishers
+  for (auto & pair : publishers_) {
+    if (can_communicate(pair.second, subscriptions_[id])) {
+      insert_sub_id_for_pub(id, pair.first, subscriptions_[id].use_take_shared_method);
+    }
+  }
+
   return id;
 }
 
 void
 IntraProcessManager::remove_subscription(uint64_t intra_process_subscription_id)
 {
-  impl_->remove_subscription(intra_process_subscription_id);
+  std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+  subscriptions_.erase(intra_process_subscription_id);
+
+  for (auto & pair : pub_to_subs_) {
+    pair.second.take_shared_subscriptions.erase(
+      std::remove(
+        pair.second.take_shared_subscriptions.begin(),
+        pair.second.take_shared_subscriptions.end(),
+        intra_process_subscription_id),
+      pair.second.take_shared_subscriptions.end());
+
+    pair.second.take_ownership_subscriptions.erase(
+      std::remove(
+        pair.second.take_ownership_subscriptions.begin(),
+        pair.second.take_ownership_subscriptions.end(),
+        intra_process_subscription_id),
+      pair.second.take_ownership_subscriptions.end());
+  }
 }
 
 void
 IntraProcessManager::remove_publisher(uint64_t intra_process_publisher_id)
 {
-  impl_->remove_publisher(intra_process_publisher_id);
+  std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+  publishers_.erase(intra_process_publisher_id);
+  pub_to_subs_.erase(intra_process_publisher_id);
 }
 
 bool
 IntraProcessManager::matches_any_publishers(const rmw_gid_t * id) const
 {
-  return impl_->matches_any_publishers(id);
+  std::shared_lock<std::shared_timed_mutex> lock(mutex_);
+
+  for (auto & publisher_pair : publishers_) {
```

Comment:
could this loop be potentially replaced by a call to `std::find_if`?

Reply:
Perhaps? I wasn't going to ask them to change that or change it myself. I think I'll pass on this for now in the interest of getting it merged. I'll add it to my fixup list however.